Adding multi-tile metro-mpi simulation support for both verilator and vcs

buhraian opened the pull request:
merging VCS/VLT compatible code into main metro-mpi repo

Commit:
Add mpic++ as argument to verilator in sims and make run_vlt possible

Commit:
successfully compiled piton_dpi.so

guillemlp committed:
Add functions to get rank from NWES and fixed compiling/executing in sims

guillemlp committed:
Add ids to the tile

guillemlp committed:
Add missign id

Commit:
fixing issues with vcs_run

Jbalkind committed:
1 chipset + 1 tile works!
- Ran sim for longer (100K cycles)
- Changed clock passed to chipset to always be core_ref_clk

Commit:
Merge branch 'metro-mpi' of github.com:guillemlp/openpiton into metro_vcs_vlt_merge

Commit:
Merge branch 'metro-mpi' of github.com:buhraian/openpiton into metro_vcs_vlt_merge

Jbalkind committed:
Set tile/chip ids properly

Jbalkind committed:
Need to put DPI under a verilator ifdef

Commit:
vcs simulation working

Commit:
corrected clock for metro_tile and metro_chipset

Jbalkind commented on build/Makefile:
```diff
@@ -1,20 +1,20 @@
+	sims -sys=metro_tile -ariane -vlt_build && sims -sys=metro_chipset -ariane -vlt_build
+
+run_both:
+	sims -sys=metro_chipset -ariane -vlt_run hello_world.c
```
@guillemlp can we remove this or are you still using it

Jbalkind replied:
Ok I just removed it so I can merge. We can always restore it if it's really needed

Jbalkind commented on piton/tools/pli/iop/bw_lib.cc:
```diff
@@ -0,0 +1,212 @@
+// Modified by Princeton University on June 9th, 2015
```
Do we want to just make this a symbolic link for now instead of having a separate file? I can't decide about it. Either that or we could just `git mv` the file but we need to make sure all the makefiles are updated if we do that

buhraian replied:
I think having separate files the way it is right now, while slightly redundant, would be best in case some files require the .c version as dependencies or some future code needs the .c instead of .cc.

Jbalkind replied:
Need for this was removed thanks to a fix to global.h

Jbalkind commented on piton/tools/pli/iop/iob_main.cc:
```diff
@@ -85,6 +85,7 @@ void init_jbus_model_call(){
 #endif // ifndef PITON_DPI
   int   idx;
 #ifndef PITON_DPI
+  printf("Hello\n");
```
Can we get rid of this lol

Jbalkind commented on piton/tools/src/sims/manycore.config:
```diff
@@ -140,6 +140,7 @@
     -vcs_build_args=-P $DV_ROOT/tools/pli/mem/bwmem_pli.tab
     -vcs_build_args=-lsocket_pli -liob -lmem_pli
     -vcs_build_args=+rad
+    -vcs_build_args="-LDFLAGS -Wl,--no-as-needed"
```
We may want to keep this as a local change Brian?

Jbalkind commented on piton/tools/src/sims/metro_chipset.config:
```diff
@@ -138,11 +138,30 @@
+    -vcs_build_args="-LDFLAGS -Wl,--no-as-needed"
+    -config_rtl=PITON_DPI 
+    -vcs_build_args="-CFLAGS -DPITON_DPI" 
+    -vcs_build_args="-CFLAGS -I/home/brianli/utils/openmpi/include" 
```
What's the deal with these includes and linker arguments? Do we need them all? Why doesn't mpic++ just know where they are?

buhraian replied:
VCS only allows for user-defined compilers like mpic++ in some places and overrides it with g++ in some parts of the Makefile it produces resulting in errors, so the workaround is to grab the output of `mpic++ --showme` and use them as CFLAGS. `mpic++ --showme` gives you the code that the arguments the wrapper is calling on top of g++, which is why it is currently hardcoded to my directories. We could specify to users to do this step themselves?

Jbalkind replied:
could you try running it without these and sending me the log on slack/zulip? I'd like to see what the errors are

Jbalkind replied:
ok we resolved this with correctly setting `-cc mpicc` and `-cpp mpic++`

Jbalkind commented on piton/tools/src/sims/metro_chipset.config:
```diff
@@ -138,11 +138,30 @@
+    //-vcs_build_args=-P $DV_ROOT/tools/pli/iop/bwioj.tab // needed for fake_l2
+    //-vcs_build_args=-P $DV_ROOT/tools/pli/socket/bwsocket_pli.tab
+    //-vcs_build_args=-P $DV_ROOT/tools/pli/mem/bwmem_pli.tab
+    //-vcs_build_args=-lsocket_pli -liob -lmem_pli
```
At this point we probably ought to add a preprocessor macro for DPI stuff... let's leave this for now though

Jbalkind replied:
Refactor done

Jbalkind commented on piton/tools/src/sims/sims,2.0:
```diff
@@ -1538,12 +1542,14 @@ sub generic_sim_build
 	
     if ($opt{vlt_build}) {
-      $build_cmd = "verilator -j 2 -cc -trace  " ;
+      $build_cmd = "verilator -j 2 -cc -trace -build " ;
```
Can we get rid of the `-j 2` and just set that in our own local .configs?

Jbalkind commented on piton/verif/env/metro_chipset/metro_chipset.v.pyv:
```diff
@@ -34,14 +34,53 @@
+  gen_riscv_dts(devices, X_TILES*Y_TILES, sysFreq, sysFreq/128, sysFreq, os.environ['ARIANE_ROOT']+"/openpiton/bootrom/", timeStamp)
+  get_bootrom_info(devices, X_TILES*Y_TILES, sysFreq, sysFreq/128, sysFreq, os.environ['ARIANE_ROOT']+"/openpiton/bootrom/", timeStamp)
+%>
+
```
I'm wondering if we ought to remove this blob from tile.v.pyv too... hm it's probably fine for now

Jbalkind commented on piton/verif/env/metro_tile/metro_tile.v.pyv:
```diff
@@ -235,9 +344,126 @@ always #25000 spi_sys_clk = ~spi_sys_clk;                       // 20MHz
+    rank = getRank();
+    size = getSize();
+    filerank.itoa(rank);
+    $dumpfile({"metro_tile_",filerank,".vcd"});
```
Does this work as intended?

buhraian replied:
Yes, str.itoa() works as intended and I've verified it produces the correct vcd files.

Jbalkind reviewed (commented):
Comments below. Can we get rid of the piton_dpi.so in the root directory too?

Commit:
resolving comments for PR

Jbalkind committed:
Adding ifdef guard on an include to allow safe compilation of bw_lib.c

Jbalkind committed:
Changing back from bw_lib.cc to bw_lib.c

Jbalkind committed:
Forgot to change metro_tile.config too

Jbalkind committed:
Fixing manycore build with vlt - macros slightly wrong in chipset

Jbalkind committed:
Changing cross module refs for clock in ciop_iob

Jbalkind committed:
Refactoring DPI file inclusion
- Move some stuff into .config files
- Add -dpi option to sims (used in manycore vcs sim)

Jbalkind committed:
Getting rid of the makefile

Jbalkind committed:
Removing commented lines
